---
title: "Assignment Week03"
author: Bohan Zhang
output: html_document
---

## Preparation
```{r warning=FALSE}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(Rsubread)      # read alignment
  library(rtracklayer)   # read/write genomic file formats
  library(Biostrings)    # handling sequences
  library(Rfastp)        # read QC/trimming
  library(epiwraps)
})
ah <- AnnotationHub()

```

## Download the Drosophila ChIP-seq for the protein CTCF

```{r, eval=FALSE, warning=FALSE}
# not important in this case, but if downloading large files on a slow connection
# we need to increase the download timeout:
options(timeout=3600)

dir.create("raw")
download.file("https://www.encodeproject.org/files/ENCFF127RRR/@@download/ENCFF127RRR.fastq.gz", "raw/ENCFF127RRR.fastq.gz", mode="wb")
```
## From the raw data, obtaining:

### (1) bam file 
Before we align the sequencing data, we need to firstly do the quality control and trimming. 
```{r}
dir.create("rfastp.trimmed")
qc <- Rfastp::rfastp("raw/ENCFF127RRR.fastq.gz", thread = 4, overrepresentationAnalysis = TRUE,
                 outputFastq = "rfastp.trimmed/ENCFF127RRR.fastq.gz")
```

This will create an html report with before/after QC plots, as well as a trimmed
and filtered fastq file. Furthermore, the R output object (equivalent to the .json 
stats also saved by the program) can be used to generate the QC plots inside a
markdown, for example:

```{r}
Rfastp::curvePlot(qc, curve="content_curves")
```

Then we start the alignment.
In R we use Rsubread to do the alignment. Firstly we need to build a genome index for mapping.

```{r, eval=FALSE}
# we get the genome sequence from AnnotationHub
genome <- ah[["AH49674"]]
# we create a new directory that will contain the genome index
dir.create("BDGP6_genome")
# we write the genome sequence in fasta format
export(import.2bit(genome), "BDGP6_genome/genome.fasta.gz", compress=TRUE)
# we build a Rsubread index
Rsubread::buildindex("BDGP6_genome/rsubread", reference="BDGP6_genome/genome.fasta.gz")
```

Then we can do the alignment:

```{r}
dir.create("aligned")
align.stats <- Rsubread::align(index="BDGP6_genome/rsubread", type="dna",
                               readfile1="rfastp.trimmed/ENCFF127RRR_R1.fastq.gz",  #the data is not paired so we only use readfile1
                               output_file="aligned/ENCFF127RRR.bam",
                               nthreads=6, sortReadsByCoordinates=TRUE)
align.stats
```

From the report we can see 3504769 (91.8%) reads are mapped.

## (2) Peaks
we call MACS from within R, through the `MACSr` package (which wraps around macs3):
```{r}
# install MACSr, if not already done:
if(!require("MACSr",quietly=TRUE)) BiocManager::install("MACSr")

library(MACSr)
dir.create("peaks")
callpeak(tfile="aligned/ENCFF127RRR.bam",
         gsize="dm", outdir = "peaks", name = "ENCFF127RRR")
# (will take a while the first time it's run, as it installs a new environment)
```

```{r}
peaks <- rtracklayer::import("peaks/ENCFF127RRR_peaks.narrowPeak")
peaks
```

There are 5392 rows of `peaks`, so we have 5392 peaks.

```{r}
head(peaks)
```

# Report
## (1) how many reads (and what percentage) were mapped: 
3504769 (91.8%) reads are mapped.

## (2) how many peaks were found: 
5392 peaks are found.

# Plot the signal around one of the peaks that is located inside a gene
Firstly we need to annotate the peaks
```{r message=FALSE, warning = FALSE}
# BiocManager::install("GenomicFeatures")
# BiocManager::install("AnnotationDbi")
# BiocManager::install("org.Dm.eg.db")
# BiocManager::install("TxDb.Dmelanogaster.UCSC.dm6.ensGene")

library(GenomicFeatures)
library(AnnotationDbi)
library(org.Dm.eg.db)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
txdb <- TxDb.Dmelanogaster.UCSC.dm6.ensGene
genes <- genes(txdb)
peaks_copy <- peaks
seqlevelsStyle(peaks_copy) <- "UCSC" 
#don't change the seqlevel style of peaks, otherwise plot is empty
overlaps <- findOverlaps(peaks_copy, genes)
peaks_on_genes <- peaks[queryHits(overlaps)]
gene_ids <- genes$gene_id[subjectHits(overlaps)]
peaks_on_genes$gene_ensembl <- gene_ids
peaks_on_genes
```

4795 peaks are mapped on gene region. we plot the first one:

```{r}
peaks_on_genes[1]
```

we can see from the "name" that the number of the peak is 406.

```{r}
plotSignalTracks(list(ENCFF127RRR = "aligned/ENCFF127RRR.bam", peaks=peaks_on_genes), region = peaks_on_genes[1],
                 extend=2000, tracks.params=list(ylim=c(0, 30)))
```


